llvm/lib/Transforms/Utils/SCCPSolver.cpp

```diff
@@ -146,6 +146,21 @@ static bool refineInstruction(SCCPSolver &Solver,
       Inst.setNonNeg();
       Changed = true;
     }
+  } else if (TruncInst *TI = dyn_cast<TruncInst>(&Inst)) {
+    auto Range = GetRange(Inst.getOperand(0));
```

Comment:
We can avoid computing ranges if both flags are already set. Same for `OverflowingBinaryOperator`.